File: skills/writing-tests-that-can-fail/evals/recorded/2026-09-25/d-account-js/baseline-haiku-1/fixture_d_account_test.mjs
import { test } from 'node:test';
import assert from 'node:assert';
import { AccountRepo, Mailer, AccountService } from './fixture_d_account.mjs';

test('AccountService.deactivate', async (t) => {
  await t.test('deactivates an active account and sends goodbye email', () => {
    const mockRepo = {
      get: (userId) => ({ email: 'user@example.com', status: 'active' }),
      setStatus: () => {},
    };
    const mockMailer = {
      sendGoodbye: () => {},
    };

    let setStatusCalled = false;
    let setStatusUserId = null;
    let setStatusValue = null;
    mockRepo.setStatus = (userId, status) => {
      setStatusCalled = true;
      setStatusUserId = userId;
      setStatusValue = status;
    };

    let sendGoodbyeCalled = false;
    let sendGoodbyeEmail = null;
    mockMailer.sendGoodbye = (email) => {
      sendGoodbyeCalled = true;
      sendGoodbyeEmail = email;
    };

    const service = new AccountService(mockRepo, mockMailer);
    service.deactivate('user123');

    assert.strictEqual(setStatusCalled, true);
    assert.strictEqual(setStatusUserId, 'user123');
    assert.strictEqual(setStatusValue, 'inactive');
    assert.strictEqual(sendGoodbyeCalled, true);
    assert.strictEqual(sendGoodbyeEmail, 'user@example.com');
  });

  await t.test('skips deactivation if account is already inactive', () => {
    const mockRepo = {
      get: (userId) => ({ email: 'user@example.com', status: 'inactive' }),
      setStatus: () => {},
    };
    const mockMailer = {
      sendGoodbye: () => {},
    };

    let setStatusCalled = false;
    mockRepo.setStatus = () => {
      setStatusCalled = true;
    };

    let sendGoodbyeCalled = false;
    mockMailer.sendGoodbye = () => {
      sendGoodbyeCalled = true;
    };

    const service = new AccountService(mockRepo, mockMailer);
    service.deactivate('user123');

    assert.strictEqual(setStatusCalled, false);
    assert.strictEqual(sendGoodbyeCalled, false);
  });

  await t.test('passes correct userId to repo.get', () => {
    const mockRepo = {
      get: (userId) => ({ email: 'user@example.com', status: 'active' }),
      setStatus: () => {},
    };
    const mockMailer = {
      sendGoodbye: () => {},
    };

    let getCalledWith = null;
    mockRepo.get = (userId) => {
      getCalledWith = userId;
      return { email: 'user@example.com', status: 'active' };
    };

    const service = new AccountService(mockRepo, mockMailer);
    service.deactivate('user456');

    assert.strictEqual(getCalledWith, 'user456');
  });

  await t.test('uses account email from repo for goodbye message', () => {
    const mockRepo = {
      get: (userId) => ({ email: 'custom@example.com', status: 'active' }),
      setStatus: () => {},
    };
    const mockMailer = {
      sendGoodbye: () => {},
    };

    let sendGoodbyeEmail = null;
    mockMailer.sendGoodbye = (email) => {
      sendGoodbyeEmail = email;
    };

    const service = new AccountService(mockRepo, mockMailer);
    service.deactivate('user123');

    assert.strictEqual(sendGoodbyeEmail, 'custom@example.com');
  });

  await t.test('calls repo and mailer in correct order', () => {
    const callOrder = [];

    const mockRepo = {
      get: (userId) => {
        callOrder.push('get');
        return { email: 'user@example.com', status: 'active' };
      },
      setStatus: () => {
        callOrder.push('setStatus');
      },
    };
    const mockMailer = {
      sendGoodbye: () => {
        callOrder.push('sendGoodbye');
      },
    };

    const service = new AccountService(mockRepo, mockMailer);
    service.deactivate('user123');

    assert.deepStrictEqual(callOrder, ['get', 'setStatus', 'sendGoodbye']);
  });
});

test('AccountRepo', async (t) => {
  await t.test('get throws error', () => {
    const repo = new AccountRepo();
    assert.throws(() => repo.get('user123'), /real DB access/);
  });

  await t.test('setStatus throws error', () => {
    const repo = new AccountRepo();
    assert.throws(() => repo.setStatus('user123', 'inactive'), /real DB access/);
  });
});

test('Mailer', async (t) => {
  await t.test('sendGoodbye throws error', () => {
    const mailer = new Mailer();
    assert.throws(() => mailer.sendGoodbye('user@example.com'), /real email send/);
  });
});

File: evals/writing-tests-that-can-fail/recorded/2026-09-25/d-account-py/baseline-haiku-1/fixture_d_account_test.py
"""Tests for Account lifecycle service."""

from unittest.mock import Mock, call
import pytest
from fixture_d_account import Account, AccountRepo, Mailer, AccountService


@pytest.fixture
def mock_repo():
    return Mock(spec=AccountRepo)


@pytest.fixture
def mock_mailer():
    return Mock(spec=Mailer)


@pytest.fixture
def account_service(mock_repo, mock_mailer):
    return AccountService(mock_repo, mock_mailer)


class TestAccountService:
    def test_deactivate_active_account(self, account_service, mock_repo, mock_mailer):
        user_id = "user123"
        email = "user@example.com"
        account = Account(user_id=user_id, email=email, status="active")
        mock_repo.get.return_value = account

        account_service.deactivate(user_id)

        mock_repo.get.assert_called_once_with(user_id)
        mock_repo.set_status.assert_called_once_with(user_id, "inactive")
        mock_mailer.send_goodbye.assert_called_once_with(email)

    def test_deactivate_inactive_account(self, account_service, mock_repo, mock_mailer):
        user_id = "user123"
        email = "user@example.com"
        account = Account(user_id=user_id, email=email, status="inactive")
        mock_repo.get.return_value = account

        account_service.deactivate(user_id)

        mock_repo.get.assert_called_once_with(user_id)
        mock_repo.set_status.assert_not_called()
        mock_mailer.send_goodbye.assert_not_called()

    def test_deactivate_returns_none(self, account_service, mock_repo):
        account = Account(user_id="user123", email="user@example.com", status="active")
        mock_repo.get.return_value = account

        result = account_service.deactivate("user123")

        assert result is None

    def test_deactivate_calls_repo_get_first(self, account_service, mock_repo, mock_mailer):
        account = Account(user_id="user123", email="user@example.com", status="active")
        mock_repo.get.return_value = account

        account_service.deactivate("user123")

        call_order = []
        mock_repo.get.side_effect = lambda uid: (call_order.append("get"), account)[1]
        mock_repo.set_status.side_effect = lambda uid, status: call_order.append("set_status")

        account_service.deactivate("user123")

        assert call_order[0] == "get"

    def test_deactivate_with_different_user_ids(self, account_service, mock_repo, mock_mailer):
        account = Account(user_id="user456", email="another@example.com", status="active")
        mock_repo.get.return_value = account

        account_service.deactivate("user456")

        mock_repo.get.assert_called_with("user456")
        mock_repo.set_status.assert_called_with("user456", "inactive")
        mock_mailer.send_goodbye.assert_called_with("another@example.com")

    def test_deactivate_with_empty_user_id(self, account_service, mock_repo, mock_mailer):
        account = Account(user_id="", email="user@example.com", status="active")
        mock_repo.get.return_value = account

        account_service.deactivate("")

        mock_repo.get.assert_called_once_with("")
        mock_repo.set_status.assert_called_once_with("", "inactive")
        mock_mailer.send_goodbye.assert_called_once_with("user@example.com")


class TestAccount:
    def test_account_creation(self):
        account = Account(user_id="user123", email="user@example.com", status="active")

        assert account.user_id == "user123"
        assert account.email == "user@example.com"
        assert account.status == "active"

    def test_account_with_inactive_status(self):
        account = Account(user_id="user123", email="user@example.com", status="inactive")

        assert account.status == "inactive"

    def test_account_equality(self):
        account1 = Account(user_id="user123", email="user@example.com", status="active")
        account2 = Account(user_id="user123", email="user@example.com", status="active")

        assert account1 == account2

    def test_account_inequality(self):
        account1 = Account(user_id="user123", email="user@example.com", status="active")
        account2 = Account(user_id="user456", email="user@example.com", status="active")

        assert account1 != account2


class TestAccountRepo:
    def test_get_not_implemented(self):
        repo = AccountRepo()

        with pytest.raises(NotImplementedError, match="real DB access"):
            repo.get("user123")

    def test_set_status_not_implemented(self):
        repo = AccountRepo()

        with pytest.raises(NotImplementedError, match="real DB access"):
            repo.set_status("user123", "inactive")


class TestMailer:
    def test_send_goodbye_not_implemented(self):
        mailer = Mailer()

        with pytest.raises(NotImplementedError, match="real email send"):
            mailer.send_goodbye("user@example.com")
